plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XWorkspaceManager.java

```diff
@@ -256,23 +328,27 @@ protected XBuildable getIncrementalDirtyBuildable(List<URI> dirtyFiles, List<URI
 	 * Generation of output files is only triggered if no source files contain unsaved changes (so-called dirty files).
 	 */
 	protected XBuildable tryIncrementalGenerateBuildable(List<URI> dirtyFiles, List<URI> deletedFiles) {
-		boolean hasSomeDirtyFiles = false;
-		for (XDocument doc : openDocuments.values()) {
-			if (doc.isDirty()) {
-				hasSomeDirtyFiles = true;
-				break;
-			}
-		}
-		if (hasSomeDirtyFiles) {
+		if (isDirty()) {
+			addFilesAwaitingGeneration(dirtyFiles, deletedFiles);
 			return getIncrementalDirtyBuildable(dirtyFiles, deletedFiles);
 		} else {
 			return getIncrementalGenerateBuildable(dirtyFiles, deletedFiles);
 		}
 	}
 
-	/** Triggers an incremental build, and will generate output files. */
+	/**
+	 * Triggers an incremental build, and will generate output files. In addition to the given files, also all
+	 * {@link #filesAwaitingGeneration files awaiting generation} will be built.
+	 */
 	protected XBuildable getIncrementalGenerateBuildable(List<URI> dirtyFiles, List<URI> deletedFiles) {
-		XBuildManager.XBuildable buildable = buildManager.getIncrementalGenerateBuildable(dirtyFiles, deletedFiles);
+		addFilesAwaitingGeneration(dirtyFiles, deletedFiles);
+
+		Pair<List<URI>, List<URI>> changedAndDeleted = filesAwaitingGeneration.getAndClear();
```

Comment:
Pair with getKey and getValue does not really express the intent of this data structure.
Maybe offer `FilesAwaitingGeneration.snapshot(): FilesAwaitingGeneration` that returns a snapshot of the original mutable structure?

Reply:
I was thinking about that, too, but did not like the fact that (1) the snapshot is mutable and, more importantly, that (2) I'd have to add two getters to class `FilesAwaitingGeneration` that are only useful in the snapshot-case, and even dangerous to use in the non-snapshot case.

Anyways, thanks to your idea of moving the entire handling to XBuildManager this code is no longer necessary.